binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java

```diff
@@ -0,0 +1,114 @@
+package io.grpc.binder.internal;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import java.util.HashSet;
+import javax.annotation.concurrent.GuardedBy;
+import io.grpc.Attributes;
+import io.grpc.Metadata;
+import io.grpc.internal.ServerStream;
+import io.grpc.internal.ServerTransportListener;
+
+/**
+ * Tracks which {@link BinderTransport.BinderServerTransport} are currently active and allows
+ * invoking a {@link BinderTransportSecurity.TerminationListener} only once all transports are
+ * terminated.
+ */
+final class ActiveTransportTracker {
+  @GuardedBy("this")
+  private final HashSet<BinderTransport.BinderServerTransport> activeTransports = new HashSet<>();
+  private final BinderTransportSecurity.TerminationListener transportSecurityTerminationListener;
+  @GuardedBy("this")
+  private boolean awaitingNoActiveTransports = false;
+  @GuardedBy("this")
+  private boolean terminationNotificationRequested = false;
+
+  /**
+   * @param transportSecurityTerminationListener invoked only once scheduled via
+   * {@link #scheduleTerminationNotification()} AND the last active transport is terminated.
+   */
+  ActiveTransportTracker(
+      BinderTransportSecurity.TerminationListener transportSecurityTerminationListener) {
+    this.transportSecurityTerminationListener = transportSecurityTerminationListener;
+  }
+
+  /**
+   * @throws IllegalStateException if {@link #scheduleTerminationNotification()} has already been
+   * called.
+   */
+  ServerTransportListener track(
+      ServerTransportListener listener,
+      BinderTransport.BinderServerTransport binderServerTransport) {
+    synchronized (this) {
+      checkState(
```

Comment:
So I believe this can actually happen today: A SETUP_TRANSPORT binder transaction can arrive while some other thread is calling BinderServer#shutdown(). Both shutdown() and handleTransaction() are synchronized so the latter method call on the binder thread will block. The implementation of BinderServer#shutdown() calls hostServiceBinder.detach() to drop any new transactions but the SETUP_TRANSPORT one in-progress has already started being handled. After shutdown() returns, handleTransaction() runs and would throw ISE here.

Instead, can you change BinderServer#handleTransaction() to handle this race by simply ignoring the SETUP_TRANSPORT transaction `if (shutdown)`? This could be a small pre-PR.

Reply:
Ack. Will start working on a separate PR in parallel.

Reply:
Cool. Let's get that one it before merging this so we don't risk spurious ISEs

Reply:
Tackling it in #11253.